From 9e1ddf6cd7ecbfd5135a0e45fbdd6fba756590f8 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 14 Sep 2021 15:09:16 +0300
Subject: [PATCH] [SW-0] arc: disable alignment checks on data memory

Currently, misaligned data access exception is observed on ARC while
running preboot which cause preboot to hang in  waiting for commands
routine. The exception is caused when address accessed by ARC is not an
integer multiple of operand size. This patch add changes to disable
alignment checks on data memory by setting AD bit to 1 in STATUS32
register.

Change-Id: I926b8e97f6541fb5cdc04a20b1e739efcc50d9b8
---

diff --git a/app/ppboot/crt0_ppboot.S b/app/ppboot/crt0_ppboot.S
index 7b6966e..c9dfb71 100644
--- a/app/ppboot/crt0_ppboot.S
+++ b/app/ppboot/crt0_ppboot.S
@@ -17,6 +17,11 @@
 .global	lf2flt
 lf2flt:
 
+	/* Disable un-aligned checks on data memory */
+	lr r0, [STATUS32]
+	bset r0, r0, 0x13
+	kflag r0
+
 	/* Clear BSS */
 	mov	r5, __bss_start
 	sub	r6, __bss_end, r5
diff --git a/app/preboot/crt0_preboot.S b/app/preboot/crt0_preboot.S
index 8c038b1..76bef06 100644
--- a/app/preboot/crt0_preboot.S
+++ b/app/preboot/crt0_preboot.S
@@ -20,6 +20,11 @@
 	mov	sp, __stack_top
 	mov	fp, sp
 
+	/* Disable un-aligned checks on data memory */
+	lr r0, [STATUS32]
+	bset r0, r0, 0x13
+	kflag r0
+
 	/* Clear BSS and stack areas */
 	mov	r5, __bss_start
 	sub	r6, __stack_top, r5
